WIP: Faster implementation of linreg

gabrielgellner opened the pull request:
Move from using the general matrix division operator to an explicit loop to calculate the OLS without extra memory allocation.

gabrielgellner commented:
I have changed the implementation of linreg following the discussion with @andreasnoack on julia-users. I would appreciate any feedback on the implementation, including what is the general preference for the return type, currently I am returning an Array but @andreasnoack  has indicated that a Tuple might be preferred for potential speed increase (which I currently can't find evidence of) as well as being a more natural output, as the user might prefer to only use one of the two returned coeffcients. I like being able to do vector arithmetic on the output, but the longer I think about it, the more agnostic I am becoming.

I have also tried to ensure that this pull request is rebased and all the other good git etiquette, but I am not sure I have done it correctly. Tips, feedback appreciated.

andreasnoack commented:
Thanks for the non-`\` version. I think it would be best to use `cov` even though it's not as fast as it should be. The reason is partly that we shouldn't reimplement an existing function. Instead, we should make the slow `cov` faster. Maybe, we should open an issue about the speed of `cov`. When it gets fixed then `linreg` will automatically benefit.

The other reason is that `cov` and `var` should (hopefully) handle all kinds of input well. E.g. your implementation will promote the result to `Float64` even for `Float32` input and it can be a bit painful to get the accumulator in the loop intialized in a generic way that allows for general input.

gabrielgellner commented:
I can easily move back to a `cov/var` version, though I wonder if it is worth it, a lot of the speedup feels like it stems from taking advantage of combining operations that are shared between `var` and `cov` during a single traversal of the data. Do you think this is just a result of `cov` being slow? 

Wouldn't having to use two functions for this lead to having the extra traversal of `x` and `y` (once for calculating cov and once for calculating var) no matter how fast the internal algorithm is? Is there a way to combine the inner iteration between the two functions they way my algorithm is across function boundaries?

Your point about the type promotion is very interesting. I hadn't thought of that. I am struck by the difficulty of doing this correctly, especially if, like both my proposals allow for using `UnitRange`/`LinRange` objects. I am not sure what to do about this. Feels like a tonne of special casing would be needed!

Finally do you have any final insight on why the `\\` version versus the `cov/var` version should be fundamentally more efficient? Maybe this is all, like you mention, over optimizing a specific case when instead `\\` should be made faster instead?

StefanKarpinski commented on .gitignore:
```diff
@@ -24,3 +24,4 @@
 *.ji
 
 .DS_Store
+get_toolchain.log
```
I don't think this should be in here.

gabrielgellner replied:
I agree. I can not for the life of me figure out how to remove that file from the pull request (or requests as it seems to have combined both). I really just want the changes to `linalg/generics.jl` and `tests/linalg/generics.jl`. But man every time I try to follow the guides on rebasing or amending ... it nukes everything. Hardest 10 line change to code I have had to do. I am might be to dumb to use git :(

tkelman replied:
I'm not opposed to adding this file to gitignore, it comes from some windows-specific setup scripts. It could maybe be a separate commit or PR but rebasing it out can be slightly complicated.

gabrielgellner replied:
So so complicated. I was about to give up programming and go back to a slide rule...

KristofferC commented on base/linalg/generic.jl:
```diff
@@ -417,7 +417,49 @@ istriu(x::Number) = true
+    tot_dx = 0.0
+    tot_y = 0.0
+    tot_xy = 0.0
+    for i in 1:n
```
can use `@inbounds` since you checked the length

gabrielgellner replied:
Do I just put the `@inbounds` in front of the for loop? Doesn't seem to change any of my timings, but seems harmless.

gabrielgellner committed:
Update linreg to faster algorithm and widen acceptable input types

Return a tuple instead of an array

Updated and added tests to deal with the tuple return change and argument types.

Ensure new linreg docstring plays nice with doc system

Moved documentation out of `helpdb` into docstring.

Remove weighted linreg from Base.

gabrielgellner commented:
Okay. I have cleaned up all the nonsense from the previous pulls. No more merge garbage, or messing with the upstream .gitignore.

I will wait a bit to see if there are anymore opinions on the two versions of the code. And if not I will make the changes suggested by @andreasnoack as he has a better feeling for the statistical functions in Base.

ararslan commented on base/linalg/generic.jl:
```diff
@@ -417,48 +417,38 @@ istriu(x::Number) = true
 Output:
 
-- `[b0, b1]` - the coefficients for the line of best fit such that yhat = b0 + b1*x
+- `(a, b)` - the coefficients for the line of best fit satisfying a + b*x
```
I actually prefer the previous wording and use of `b0` and `b1`. If it were me, I would just change the square brackets to parentheses and leave the description unchanged.

gabrielgellner replied:
I am happy to use either, see if there is any consensus. I changed this so that it was consistent with the example given in the `Base.rst` file. Changing to the parentheses is meant to be consistent with the code soon, I just can't figure out how to use the `@test_approx_equal_eps` with a tuple return instead of the array...

ararslan commented on base/linalg/generic.jl:
```diff
@@ -417,48 +417,38 @@ istriu(x::Number) = true
+function linreg(x::AbstractVector, y::AbstractVector)
     # Least squares given
-    # Y = b0 + b1*X
+    # Y = a + b*X
```
Out of curiosity, why change to `a` and `b`? Personally I've always found the subscripted beta notation for regression coefficients much clearer than `a` and `b`.

gabrielgellner replied:
Yeah my brain clearly gravitates to the beta notation, but when looking over the existing code I realized I had introduced inconsistency with other parts of the code so I changed this.

ararslan commented on base/linalg/generic.jl:
```diff
@@ -417,48 +417,38 @@ istriu(x::Number) = true
-    return [b0, b1]
+    # b = cov(X, Y)/var(X)
+    # a = mean(Y) - b*mean(X)
+    size(x) == size(y) || throw(DimensionMismatch())
```
It might be nice to provide at least a minimal decription in the `DimensionMismatch` exception constructor.

gabrielgellner replied:
I will look into this. Was just following the code from the recent changes to `quantile`

ararslan commented on test/linalg/generic.jl:
```diff
@@ -65,26 +65,34 @@ let X = [3  9   5;
 @test_approx_eq linreg(x,y) [2.5559090909090867, 1.6960139860139862]
 @test_approx_eq linreg(sub(x,1:6),sub(y,1:6)) [3.8366666666666642,1.3271428571428574]
 
-##The following tests will not require collect once issue #16520 is fixed
```
Has that been fixed? The issue is still open.

gabrielgellner replied:
The issue is still a problem, but since this code no longer uses the `\` operator it is not effected and the tests work as intended, hence the removal of the `collect` hacks.

andreasnoack commented on test/linalg/generic.jl:
```diff
@@ -65,11 +65,52 @@ let X = [3  9   5;
+# check (LinSpace, UnitRange)
+x = linspace(1.0, 12.0, 100)
+y = -100:-1
+@test_approx_eq linreg(x, y) [-109.0, 9.0]
```
In case you return a tuple, you can do `[linreg(x,y)...]` to convert it to an array here.

gabrielgellner replied:
Great thanks. I have changed the return type to a tuple as you have suggested, and the tests to reflect this.

andreasnoack commented:
LGTM. Do you consider it ready? If so then please [squash the commits](https://github.com/ginatrapani/todo.txt-android/wiki/Squash-All-Commits-Related-to-a-Single-Issue-into-a-Single-Commit) (migth be little fight with git the first time you do it but you'll have to go through this process sooner or later).

KristofferC commented:
Why not squash with github ui?

tkelman commented:
Is this docstring intended to replace the old one? Don't think that's ready yet, should run `make docs` to check that it's propagating into the rst.

andreasnoack commented:
@KristofferC I didn't know about that option. So if an owner merges the PR, he can do the squash?

KristofferC commented:
Yes, when you press merge you can chose to squash it (and edit the commit message if you chose). It will also avoid the merge commit.

tkelman commented:
See https://github.com/blog/2141-squash-your-commits - when you click "Merge pull request," but before you click "Confirm merge," there's an option you can select. It's sticky though, so keep an eye on it since you might not want to squash commits from every single PR, for those that are broken into good separate (all passing) pieces that may be worth bisecting within later.

gabrielgellner commented:
@tkelman There was no docstring before, but there is an entry in `base/docs/helpdb/Base.jl` which I have not changed. I can run the `make docs`, will this create a new file I need to commit?

@andreasnoack I am happy to try and figure out how to do the squash if that is preferred ... learned more about git then I ever wanted in the last couple of days ...

tkelman commented:
The contents of `base/docs/helpdb/Base.jl` are docstrings, they're just all in one leftover file from the doc migration last year. They're supposed to gradually get moved inline to docstrings at the function signature like the new one you're adding. Having a docstring in `helpdb/Base.jl` and inline for the same function is bad, I'm not sure which will show up when you query the repl help - they might both get concatenated which would be redundant here.

So I'd recommend deleting the one from `helpdb/Base.jl`, maybe incorporating anything that's worth keeping into your new docstring. Then you run `make docs` and it should translate your markdown inline docstring into rst and populate it in the "generated from Julia source" section of the rst manual. Commit that change and add it here, so the next person who runs `make docs` doesn't get confused by your change which is unrelated to whatever they're working on. It's a messy system but we're aiming to eventually move away from rst and sphinx altogether. Next release cycle, hopefully.

gabrielgellner commented:
@tkelman okay I have merged the docstring and the information in `helpdb`. Before I push this to get feedback are there any guidelines for docstrings? I have followed the example given in `linalg/generics.jl:normalize` which seem to follow the scipy format, but when I look elsewhere this does not seem to be followed. Is there a preferred layout or is it function to function?

@andreasnoack fixing up the docstrings I see the case for weighted `linreg`, do we want to change the algorithm on this or do you think it is fine as is?

tkelman commented:
ref https://github.com/JuliaLang/julia/pull/15136 for the most recent discussion on docstring guidelines

andreasnoack commented:
I'd forgotten the weighted version. Actually, it shouldn't be here. We have moved all the weighted statistics functions to `StatsBase.jl` so that is where it belongs. In `StatsBase.jl`, we could also simply use the weighted variance and covariance functions there. Would you mind opening a PR against `StatsBase.jl` with a similar version

gabrielgellner commented:
@tkelman I am not sure what is happening but the `rst` file is not being updated when I do `make docs` I have tried doing a `make clean` first but the section in `doc/stdlib/linalg.rst` is not reflecting the changes I have made to the docstrings and is still reflecting the deleted section in `helpdb` as well as having a information from the deleted weighted `linreg` method. Not sure what I am doing wrong. The help from the REPL is all good now.

@andreasnoack I have removed the weighted version, once I get the documentation stuff fixed up I will open a pull request in `StatsBase.jl`. Thanks for all the feedback.

gabrielgellner commented:
Okay I believe everything is good to go! Let me know if all the recent changes look good and I can figure out how to squash the commits.

ararslan commented:
`git reset --soft HEAD~X` where `X` is the number of commits you want to squash (i.e. squash the `X` most recent commits into one), then `git commit` and write a new commit message. :smile:

KristofferC commented:
I always do `git rebase -i HEAD~X` and shose squash

gabrielgellner commented:
So I have my local copy of the branch rebased and squashed. But I don't understand how to push it over this current pull request.

I tried
git push -f origin

but I get strange error messages about no the name of the upstream branch. Am I missing something simple?

tkelman commented:
You need to force push over the `linreg-updates` branch of the remote that corresponds to your fork. Provide both the remote name and the branch name to push.

gabrielgellner commented:
OMG I hate git. Disregard all the commit noise ...

gabrielgellner commented:
Okay ready to rock.

tkelman commented on base/linalg/generic.jl:
```diff
@@ -417,12 +417,40 @@ istriu(x::Number) = true
+"""
+    linreg(x, y)
+
+Perform simple linear regression using Ordinary Least Squares. Return `a` and `b` such
```
Returns

andreasnoack commented:
Thanks for the contribution.
